src/decompress.cpp

```diff
@@ -29,4 +30,54 @@ auto read_header(huffman::bit_span& compressed_bits)
   compressed_bits.consume(3);
   return BlockHeader{final, type};
 }
-}  // namespace starflate::detail
+
+}  // namespace detail
+
+auto decompress(std::span<const std::byte> src, std::span<std::byte> dst)
+    -> std::expected<DecompressResult, DecompressError>
+{
+  using enum detail::BlockType;
+
+  huffman::bit_span src_bits{src};
+  while (true) {
+    const auto header = detail::read_header(src_bits);
+    if (not header) {
+      return std::unexpected{header.error()};
+    }
+    if (header->type == NoCompression) {  // no compression
+      // Any bits of input up to the next byte boundary are ignored.
+      src_bits.consume_to_byte_boundary();
+      const std::uint16_t len = src_bits.pop_16();
+      const std::uint16_t nlen = src_bits.pop_16();
+      if (len != static_cast<std::uint16_t>(~nlen)) {
+        return std::unexpected{DecompressError::NoCompressionLenMismatch};
+      }
+      // TODO: should we return an error instead of assert?
+      assert(
+          std::cmp_greater_equal(
+              src_bits.size(), std::size_t{len} * CHAR_BIT) and
+          "not enough bits in src");
+
+      if (std::ranges::size(dst) < len) {
+        return DecompressResult{src, dst};
```

Comment:
Related to the above comment on return information. Here the output buffer needs to be expanded before calling `decompress` again. Can we propagate the missing number of bytes that's needed in the output buffer?